File: Simple-VA-Poisson/PoisReg.R
library(TMB)
local({
  dir <- "Simple-VA-Poisson"
  if(!grepl(paste0(dir, "$"), getwd()))
    setwd(dir)
})

# compile code
stopifnot(compile("PoisReg.cpp") == 0)
dyn.load(dynlib("PoisReg"))

# simulate date
set.seed(42163306)
dat_org <- within(list(), {
  sigma <- .5
  beta <- c(2, 1.5)
  grp_size <- 5L
  n_groups <- 500L
  
  x <- runif(grp_size * n_groups, -1, 1)
  grp <- rep(1:n_groups, each = grp_size)
  y <- rpois(grp_size * n_groups, lambda = exp(
    beta[1] + beta[2] * x + rnorm(n_groups, sd = sigma)[grp]))
})

#####
# setup functions to use to optimize variational parameters and parameters 
# in the model
params <- with(dat_org, list(
  sigma_log = 0,
  beta = c(log(mean(y)), 0),
  mu = rep(0, n_groups),
  lambda_log = rep(0, n_groups)))
ad_func <- MakeADFun(
  data = dat_org[c("y", "x", "grp_size")], parameters = params, DLL = "PoisReg",
  silent = TRUE)

# function that returns a function to optimize a subset of the parameters 
get_top_func <- function(params_update, reltol = 1e-5){
  eval(bquote(function(obj, verbose = FALSE){
    # setup the objects we need
    par_all <- obj$par
    idx_update <- names(par_all) %in% .(params_update)
    par <- par_all[idx_update]
    fn <- obj$fn
    gr <- obj$gr
    
    # setup functions we need
    fn_use <- function(x, ...){
      z <- par_all
      z[idx_update] <- x
      fn(z)
    }
    gr_use <- function(x, ...){
      z <- par_all
      z[idx_update] <- x
      drop(gr(z))[idx_update]
    }
    
    # optimize
    opt_res <- optim(par, fn = fn_use, gr = gr_use, method = obj$method, 
                     control = list(fnscale = -1, maxit = 1000L, 
                                    reltol = .(reltol)))
    if(opt_res$convergence > 0)
      stop(sprintf("optim failed with code %d\n", opt_res$convergence))
    obj$par[idx_update] <- opt_res$par
    obj$value <- opt_res$value
    
    # print and return
    if(verbose){
      cat(sprintf("\nLower bound is %.4f. Parameters estimates are:\n", 
                  obj$value))
      print(obj$par[idx_update])
    }
    
    obj
  }))
}

#####
# use function with two-step procedure where we first optimize over variational 
# parameters and then over model parameters
optim_two_step <- function(obj, maxit = 1000L, eps = 1e-8, verbose = TRUE){
  opt_variational <- get_top_func(c("mu", "lambda_log"), reltol = eps)
  opt_params <- get_top_func(c("sigma_log", "beta"), reltol = eps)
  
  lls <- rep(NA_real_, maxit)
  ll_old <- -.Machine$double.xmax
  for(i in 1:maxit){
    obj <- opt_variational(obj, verbose = FALSE)
    obj <- opt_params(obj, verbose = verbose && (i - 1L) %% 10 == 0)
    
    val <- obj$value
    lls[i] <- val
    if(abs((val - ll_old) / ll_old) < eps)
      break
    ll_old <- val
  
  }
  
  obj$n_it <- i
  obj$lls <- lls[1:i]
  obj
  
}

two_est <- optim_two_step(ad_func)

plot(two_est$lls) # lower bounds vs iterations
exp(two_est$par["sigma_log"]) # estimated std. dev.

#####
# just maximize directly 
ad_func$control <- list(fnscale = -1, maxit = 1000L)
one_est <- do.call(optim, ad_func)

# difference in parameter estimates
two_est$par[1:3] - one_est$par[1:3]
one_est$par[1:3]
exp(one_est$par["sigma_log"])

#####
# compare w/ adaptive gaussian quadrature
library(lme4)
lme4_fit <- glmer(y ~ x + (1 | grp), dat_org, poisson(), nAGQ = 10)
summary(lme4_fit)

# compare log-likelihood and lower bound
logLik(lme4_fit)
one_est$value

File: Simple-VA-Poisson/PoisReg.cpp
#include <TMB.hpp>
#include <stdexcept>

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(y);
  DATA_VECTOR(x);
  DATA_INTEGER(grp_size);
  
  PARAMETER(sigma_log);
  PARAMETER_VECTOR(beta);
  PARAMETER_VECTOR(mu);
  PARAMETER_VECTOR(lambda_log);
  
  Type sigma = exp(sigma_log);
  vector<Type> lambda = exp(lambda_log);
  
  /* check args. The code currently crash with C++ exceptions though... */
  std::size_t const n = y.size(), n_groups = n / grp_size;
  if(n != (unsigned)x.size() or n < (unsigned)grp_size or n % grp_size > 0)
    throw std::invalid_argument("sizes differ");
  if(n_groups != (unsigned)mu.size() or n_groups != (unsigned)lambda.size())
    throw std::invalid_argument("invalid group size");
  
  /* compute lower bound */
  Type lb = Type(0);
  {
    Type const sigma_sq = sigma * sigma, 
                   half(.5);
    lb += Type(n_groups) * half * (1. - log(sigma_sq));
    
    std::size_t j = 0;
    for(std::size_t i = 0; i < n_groups; ++i){
      Type const 
        lambda_half = lambda[i] * half,
               term = mu[i] + lambda_half;
      lb += half * (log(lambda[i]) - (mu[i] * mu[i] + lambda[i]) / sigma_sq);
      
      std::size_t const n_j = (i + 1) * grp_size;
      for(; j < n_j; ++j)
        lb += dpois(y[j], exp(beta[0] + beta[1] * x[j] + term), true) - 
            y[j] * lambda_half;
    }
  }
  
  return lb;
}
